src/include/detail/graph/vamana.h

```diff
@@ -0,0 +1,1019 @@
+   * (j,N_"out " (j),α,R) to update out-neighbors of j.
+   */
+  template <class Array>
+  void train(const Array& training_set) {
+    feature_vectors_ = std::move(ColMajorMatrix<feature_type>(
+        training_set.num_rows(), training_set.num_cols()));
+    std::copy(
+        training_set.data(),
+        training_set.data() + training_set.num_rows() * training_set.num_cols(),
+        feature_vectors_.data());
+
+    dimension_ = feature_vectors_.num_rows();
+    num_vectors_ = feature_vectors_.num_cols();
+    // graph_ = ::detail::graph::init_random_adj_list<feature_type, id_type>(
+    //     feature_vectors_, R_max_degree_);
+
+    graph_ = ::detail::graph::adj_list<feature_type, id_type>(num_vectors_);
+    // dump_edgelist("edges_" + std::to_string(0) + ".txt", graph_);
+
+    medioid_ = medioid(feature_vectors_);
+
+    debug_index();
+    size_t counter{0};
+    //    for (float alpha : {alpha_min_, alpha_max_}) {
+    // Just use one value of alpha
+    for (float alpha : {alpha_max_}) {
+      scoped_timer _("train " + std::to_string(counter), true);
+      size_t total_visited{0};
+      for (size_t p = 0; p < num_vectors_; ++p) {
+        ++counter;
+
+        // Do not need top_k or top_k scores here -- use path_only enum
+        auto&& [top_k_scores, top_k, visited] = greedy_search(
+            graph_,
+            feature_vectors_,
+            medioid_,
+            feature_vectors_[p],
+            1,
+            L_build_,
+            sum_of_squares_distance{});
+        total_visited += visited.size();
+
+        robust_prune(
+            graph_,
+            feature_vectors_,
+            p,
+            visited,
+            alpha,
+            R_max_degree_,
+            sum_of_squares_distance{});
+        {
+          scoped_timer _{"post search prune"};
+          for (auto&& [i, j] : graph_.out_edges(p)) {
+            // @todo Do this without copying -- prune should take vector of
+            //  tuples and p (it copies anyway) maybe scan for p and then only
+            //  build tmp after if?
+            auto tmp = std::vector<size_t>(graph_.out_degree(j) + 1);
+            tmp.push_back(p);
+            for (auto&& [_, k] : graph_.out_edges(j)) {
+              tmp.push_back(k);
+            }
+
+            if (size(tmp) > R_max_degree_) {
+              robust_prune(
+                  graph_,
+                  feature_vectors_,
+                  j,
+                  tmp,
+                  alpha,
+                  R_max_degree_,
+                  sum_of_squares_distance{});
+            } else {
+              graph_.add_edge(
+                  j,
+                  p,
+                  sum_of_squares_distance()(
+                      feature_vectors_[p], feature_vectors_[j]));
+            }
+          }
+        }
+        if ((counter) % 10 == 0) {
+          // dump_edgelist("edges_" + std::to_string(counter) + ".txt", graph_);
+        }
```

Comment:
(Minor) Let's remove this? Or if you'd like to keep it, would suggest something like:
```
        // Use this to watch the graph change over time:
        // if ((counter) % 10 == 0) {
          // dump_edgelist("edges_" + std::to_string(counter) + ".txt", graph_);
        // }
```